File: languages/ruby/lib/benchmarkers/benchmarker.rb
# Implements base benchmarking logic

class Benchmarker
  NANO_PER_MILLI = 1_000_000

  def initialize(operation_name)
    @operation_name = operation_name
  end

  def consume_file_input(input_file_path)
    raise NotImplementedError("Expected to be implemented by child class!")
  end
  
  def get_operation_execution_time()
    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
    yield
    end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)

    end_time - start_time
  end

  def get_operation_execution_results(execution_count, &operation_execution_block)
    [*0..execution_count-1].map { get_operation_execution_time }
  end

  def print_benchmark_analysis(execution_count)
    results = get_operation_execution_results(execution_count)
    total_time = results.sum

    milliseconds = total_time.to_f / NANO_PER_MILLI
    formatted_time = sprintf("%.6f", milliseconds)

    puts "Ruby's #{@operation_name} execution time (over #{execution_count} loops): #{formatted_time} ms"
  end
end

File: languages/python/src/benchmarkers/benchmarker.py
from abc import ABC, abstractmethod
import time

class Benchmarker(ABC):
    @abstractmethod
    def __init__(self, input_file_path: str):
        self.operation_name = "Unknown Operation"

    @abstractmethod
    def _get_input_data_args(self) -> list:
        """
        Gets list of input parameters that will be passed to the operation being
        performed and having its execution time measured. This getter method must
        be idempotent, so any non-primitive types should only have copies of the
        original data returned.
        """
        ...

    @abstractmethod
    def _perform_operation(self, *args):
        """
        Performs operation defined by the child class that inherits this base class.
        """
        ...

    def _get_operation_execution_time(self):
        """
        Performs the operation and returns the execution time of it.

        Args:
            verify (bool): If True, the operation's results will be verified to ensure
            that its output is correct. If it returns invalid output, an Exception will
            be raised.

        returns:
            float: The execution time of the operation in seconds.
        """

        input_data_args = self._get_input_data_args()

        # Performs operation and calculates execution time
        start_time = time.time()
        self._perform_operation(*input_data_args)
        end_time = time.time()

        return end_time - start_time

    def print_benchmark_analysis(self, execution_count=1000):
        """
        Prints the execution time of the operation being benchmarked after executing
        it the specified number of times.

        Args:
            execution_count (int): The number of times the benchmarked operation is being executed.
        """

        total_time = sum(self._get_operation_execution_time() for _ in range(execution_count))

        # Prints statistics
        print(f"Python's {self.operation_name} execution time (over {execution_count} loops): {round(total_time * 1000, 6)} ms")